### PR TITLE
Refactor non-linear solvers

### PR DESCRIPTION
**Pull request by devmotion**
This PR creates separate cache types for each non-linear solver which for Anderson acceleration contains now also caches `alphas` (only for in-place version) and `residuals`.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
I'll look in more detail, it does look good. However, I do want to look into https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/647 before we really dig deeper into nonlinear solver changes.

File: src/caches/adams_bashforth_moulton_caches.jl [hunk 1]
```diff
@@ -979,29 +978,28 @@ end
...
+  @iipnlsolve
+
   k1 = zero(rate_prototype)
   k2 = zero(rate_prototype)
   du₁ = zero(rate_prototype)
-
   uprev2 = similar(u)
   uprev3 = similar(u)
   tprev2 = t
+  atmp = similar(u,uEltypeNoUnits)
```
Comment (ChrisRackauckas) [inline src/caches/adams_bashforth_moulton_caches.jl]: this method isn't adaptive?

File: src/caches/adams_bashforth_moulton_caches.jl [hunk 2]
```diff
@@ -917,28 +917,27 @@ end
...
+  γ, c = 1//2, 1
+  @iipnlsolve
+
   k1 = zero(rate_prototype)
   k2 = zero(rate_prototype)
   du₁ = zero(rate_prototype)
-
   uprev3 = similar(u)
   tprev2 = t
+  atmp = similar(u,uEltypeNoUnits)
```
Comment (ChrisRackauckas) [inline src/caches/adams_bashforth_moulton_caches.jl]: also not adaptive?

File: src/caches/bdf_caches.jl [hunk 3]
```diff
@@ -106,37 +111,35 @@ end
...
+
   k₁ = similar(rate_prototype)
   k₂ = order >= 3 ? zero(rate_prototype) : k₁
   k₃ = order == 4 ? zero(rate_prototype) : k₁
   du₁ = zero(rate_prototype)
   du₂ = zero(rate_prototype)
 
-  γ, c = 1//1, 1
-  @iipnlsolve
+  uprev2 = similar(u)
```
Comment (ChrisRackauckas) [inline src/caches/bdf_caches.jl]: make these zero

File: src/caches/bdf_caches.jl [hunk 4]
```diff
@@ -422,11 +425,12 @@ end
 
 function alg_cache(alg::MEBDF2,u,rate_prototype,uEltypeNoUnits,uBottomEltypeNoUnits,
                    tTypeNoUnits,uprev,uprev2,f,t,dt,reltol,p,calck,::Type{Val{true}})
-  @iipnlcachefields
-  z₁ = similar(u); z₂ = similar(u); z₃ = similar(u); tmp2 = similar(u);
-  atmp = similar(u,uEltypeNoUnits)
   γ, c = 1, 1
   @iipnlsolve
+
+  z₁ = similar(u); z₂ = similar(u); z₃ = similar(u); tmp2 = similar(u)
```
Comment (ChrisRackauckas) [inline src/caches/bdf_caches.jl]: make these zero

File: src/caches/kencarp_kvaerno_caches.jl [hunk 5]
```diff
@@ -56,9 +54,9 @@ function alg_cache(alg::KenCarp3,u,rate_prototype,uEltypeNoUnits,uBottomEltypeNo
     k3 = nothing; k4 = nothing
     uf = DiffEqDiffTools.UJacobianWrapper(f,t,p)
   end
-  tab = KenCarp3Tableau(uToltype,real(tTypeNoUnits))
-  γ, c = tab.γ, tab.c3
-  @iipnlsolve
+
+  z₁ = similar(u); z₂ = similar(u); z₃ = similar(u); z₄ = z
```
Comment (ChrisRackauckas) [inline src/caches/kencarp_kvaerno_caches.jl]: zero

File: src/caches/kencarp_kvaerno_caches.jl [hunk 6]
```diff
@@ -109,15 +106,13 @@ end
...
-  z₁ = similar(u); z₂ = similar(u);
-  z₃ = similar(u); z₄ = similar(u)
-  z₅ = z
-  atmp = similar(u,uEltypeNoUnits)
-  tab = Kvaerno4Tableau(uToltype,real(tTypeNoUnits))
+  tab = Kvaerno4Tableau(real(uBottomEltypeNoUnits),real(tTypeNoUnits))
   γ, c = tab.γ, tab.c3
   @iipnlsolve
 
+  z₁ = similar(u); z₂ = similar(u); z₃ = similar(u); z₄ = similar(u); z₅ = z
```
Comment (ChrisRackauckas) [inline src/caches/kencarp_kvaerno_caches.jl]: zero

File: src/caches/kencarp_kvaerno_caches.jl [hunk 7]
```diff
@@ -70,16 +68,15 @@ end
...
-  tprev2 = t
-
-  tab = Kvaerno4Tableau(uToltype,real(tTypeNoUnits))
+function alg_cache(alg::Kvaerno4,u,rate_prototype,uEltypeNoUnits,uBottomEltypeNoUnits,
+                   tTypeNoUnits,uprev,uprev2,f,t,dt,reltol,p,calck,::Type{Val{false}})
+  tab = Kvaerno4Tableau(real(uBottomEltypeNoUnits),real(tTypeNoUnits))
   γ, c = tab.γ, tab.c3
   @oopnlsolve
 
+  uprev3 = u
```
Comment (ChrisRackauckas) [inline src/caches/kencarp_kvaerno_caches.jl]: why is this needed?

File: src/caches/kencarp_kvaerno_caches.jl [hunk 8]
```diff
@@ -130,13 +125,13 @@ end
...
                    uprev,uprev2,f,t,dt,reltol,p,calck,::Type{Val{false}})
-  @oopnlcachefields
-  uprev3 = u
-  tprev2 = t
-  tab = KenCarp4Tableau(uToltype,real(tTypeNoUnits))
+  tab = KenCarp4Tableau(real(uBottomEltypeNoUnits),real(tTypeNoUnits))
   γ, c = tab.γ, tab.c3
   @oopnlsolve
 
+  uprev3 = u
```
Comment (ChrisRackauckas) [inline src/caches/kencarp_kvaerno_caches.jl]: more of these?

**Comment by devmotion** [comment]
Good to know but maybe it's better to address your comments in a separate PR? This PR only changes the macros for the creation of the non-linear solvers and ensures that the non-linear solvers are created before (most) other fields of the integrator caches but does not affect the caches otherwise.
